src/Training.cpp

```diff
@@ -158,6 +158,84 @@ void Training::dump_training(int game_score, const std::string& out_filename) {
     dump_training(game_score, chunker);
 }
 
+Network::BoardPlane fix_v2(Network::BoardPlane plane) {
+    for (int i = 0, n = plane.size(); i < n; i+=8) {
+        for (auto j = 0; j < 4; j++) {
+            bool t = plane[i+j];
+            plane[i+j] = plane[i+8-j-1];
+            plane[i+8-j-1] = t;
+        }
+    }
+
+    return plane;
+}
+
+
+void Training::dump_training_v2(int game_score, OutputChunker& outchunk) {
+    /**
+     * From chunkparser.py
+     *
+     * V2 Format (8604 bytes total)
+     * int32 version (4 bytes)
+     * 1924 float32 probabilities (7696 bytes)
+     * 112 packed bit planes (896 bytes)
+     * uint8 castling us_ooo (1 byte)
+     * uint8 castling us_oo (1 byte)
+     * uint8 castling them_ooo (1 byte)
+     * uint8 castling them_oo (1 byte)
+     * uint8 side_to_move (1 byte)
+     * uint8 rule50_count (1 byte)
+     * uint8 move_count (1 byte)
+     * int8 result (1 byte)
+     * self.v2_struct = struct.Struct('4s7696s896sBBBBBBBb')
+     */
+    static int VERSION = 2;
+
+    std::stringstream out;
+    for (const auto& step : m_data) {
+        // Store the binary version number (4 bytes)
+        out.write(reinterpret_cast<char*>(&VERSION), sizeof(VERSION));
+
+        // Then the move probabilities (7696 bytes)
+        assert(step.probabilities.size()*sizeof(step.probabilities[0]) == 7696);
+        for (auto p : step.probabilities) {
+            uint32 *vp = reinterpret_cast<uint32*>(&p);
+            uint32 v = htole32(*vp);
+            out.write(reinterpret_cast<char*>(&v), sizeof(v));
+        }
+
+        // 112 bitplanes (896 bytes)
+        int kFeatureBase = Network::T_HISTORY * 14;
+        assert(kFeatureBase*sizeof(step.planes.bit[0].to_ullong()) == 896);
+        for (int p = 0; p < kFeatureBase; p++) {
+            const auto& plane = fix_v2(step.planes.bit[p]);
+            auto val = htole64(plane.to_ullong());
+            assert(plane.size() % 4 == 0);
```

Comment:
Methinks this code depends more specifically on `plane.size() == 64` (or `plane.size() == std::numeric_limits<unsigned long long>::digits`, if you want to be pedantic), because of the way it uses `to_ullong()`.

Reply:
Indeed, but the assert at line 209 implies this already. It's kind of unnecessary here I think?

Reply:
I don't think it does. The assert at line 209 is equivalent to `assert(kFeatureBase*sizeof(unsigned long long) == 896);` It doesn't imply anything about `plane.size()`.